xds/internal/xdsclient/tests/lds_watchers_test.go

```diff
@@ -919,6 +943,128 @@ func (s) TestLDSWatch_NACKError(t *testing.T) {
 	if gotErr == nil || !strings.Contains(gotErr.Error(), wantListenerNACKErr) {
 		t.Fatalf("update received with error: %v, want %q", gotErr, wantListenerNACKErr)
 	}
+
+	// Verify that the expected error is propagated to the new watcher as well.
+	lw2 := newListenerWatcher()
+	ldsCancel2 := xdsresource.WatchListener(client, ldsName, lw2)
+	defer ldsCancel2()
+	u, err = lw2.updateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout when waiting for a listener resource from the management server: %v", err)
+	}
+	gotErr = u.(listenerUpdateErrTuple).err
+	if gotErr == nil || !strings.Contains(gotErr.Error(), wantListenerNACKErr) {
```

Comment:
The error type in the `xdsresource` package provides a way to get to the underlying error type. See: https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/xdsresource/errors.go#L61

I think we should use that instead of string comparison.

I agree existing tests are not doing that. But that is not a good enough reason to make new code not do that. In fact, it would be wonderful in existing code could be cleaned up as well.

Reply:
I think for NACK error of no routing, we don't have any specific error type. So, type will also be unknown. That's we are verifying the string.

Reply:
What do you think about adding a new type for NACK errors? If you think it makes sense to do that, we should file an issue to track it and eventually get to it. Might be good to get to (and fixing tests) before making the client public.

Reply:
Filed an issue https://github.com/grpc/grpc-go/issues/7863. I think it should be simple change as i described in the issue to set the NACK error type while decoding. Though it still can be separate PR because we will have to update all the tests.